From: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Date: Thu, 12 Apr 2018 23:54:13 +0200
Subject: Unity launcher support

Last-Update: 2026-07-15
---
 meson.build                                        |   6 +
 meson_options.txt                                  |   6 +
 src/meson.build                                    |  10 ++
 src/nautilus-application.c                         |   9 ++
 src/nautilus-progress-indicator.c                  |  66 ++++++++
 src/unity-bookmarks-handler.c                      | 145 ++++++++++++++++++
 .../unity-bookmarks-handler.h                      |  17 ++-
 src/unity-quicklist-handler.c                      | 166 +++++++++++++++++++++
 src/unity-quicklist-handler.h                      |  69 +++++++++
 9 files changed, 488 insertions(+), 6 deletions(-)
 create mode 100644 src/unity-bookmarks-handler.c
 copy extensions/image-properties/nautilus-image-properties-model.h => src/unity-bookmarks-handler.h (57%)
 create mode 100644 src/unity-quicklist-handler.c
 create mode 100644 src/unity-quicklist-handler.h

diff --git a/meson.build b/meson.build
index 62b24a0..babf20c 100644
--- a/meson.build
+++ b/meson.build
@@ -133,6 +133,11 @@ if get_option('extensions')
   gst_pbutils_dep = dependency('gstreamer-pbutils-1.0')
 endif
 
+if get_option ('unity-launcher')
+  unity = dependency ('unity', version: '>=4.0.0')
+  dbusmenu = dependency ('dbusmenu-glib-0.4', version: '>=0.4.90')
+endif
+
 ####################
 # End dependencies #
 ####################
@@ -171,6 +176,7 @@ conf.set_quoted('PACKAGE_VERSION', meson.project_version())
 conf.set_quoted('PROFILE', profile)
 conf.set_quoted('VERSION', version_string)
 
+conf.set('HAVE_UNITY', get_option('unity-launcher'))
 conf.set('HAVE_SELINUX', selinux.found())
 conf.set('HAVE_CLOUDPROVIDERS', cloudproviders.found())
 
diff --git a/meson_options.txt b/meson_options.txt
index 54e9687..b3121e5 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -38,6 +38,12 @@ option(
   deprecated: {'true': 'enabled', 'false': 'disabled'},
   description: 'Enable SELinux context support in file properties dialog',
 )
+option(
+  'unity-launcher',
+  type: 'boolean',
+  value: false,
+  description: 'build with unity launcher support',
+)
 option(
   'cloudproviders',
   type: 'feature',
diff --git a/src/meson.build b/src/meson.build
index 35b5482..c37e884 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -289,6 +289,16 @@ nautilus_deps = [
   xdp_gnome,
 ]
 
+if get_option ('unity-launcher')
+  nautilus_deps += [unity, dbusmenu]
+  libnautilus_sources += [
+    'unity-quicklist-handler.c',
+    'unity-quicklist-handler.h',
+    'unity-bookmarks-handler.c',
+    'unity-bookmarks-handler.h'
+  ]
+endif
+
 libnautilus = static_library(
   'nautilus',
   libnautilus_sources,
diff --git a/src/nautilus-application.c b/src/nautilus-application.c
index c8a96dc..89a713c 100644
--- a/src/nautilus-application.c
+++ b/src/nautilus-application.c
@@ -25,6 +25,7 @@
  */
 #define G_LOG_DOMAIN "nautilus-application"
 
+#include <config.h>
 #include "nautilus-application.h"
 
 #include <adwaita.h>
@@ -69,6 +70,10 @@
 #include "nautilus-ui-utilities.h"
 #include "nautilus-window.h"
 
+#ifdef HAVE_UNITY
+#include "unity-bookmarks-handler.h"
+#endif
+
 struct _NautilusApplication
 {
     AdwApplication parent_instance;
@@ -1013,6 +1018,10 @@ nautilus_application_startup (GApplication *app)
         maybe_migrate_gtk_filechooser_preferences ();
     }
 
+#ifdef HAVE_UNITY
+    unity_bookmarks_handler_initialize ();
+#endif
+
     g_signal_connect (self, "shutdown", G_CALLBACK (on_application_shutdown), NULL);
 
     g_signal_connect_object (gtk_icon_theme_get_for_display (gdk_display_get_default ()),
diff --git a/src/nautilus-progress-indicator.c b/src/nautilus-progress-indicator.c
index 16181aa..573eb7e 100644
--- a/src/nautilus-progress-indicator.c
+++ b/src/nautilus-progress-indicator.c
@@ -13,6 +13,12 @@
 #include "nautilus-progress-paintable.h"
 #include "nautilus-window.h"
 
+#include <config.h>
+
+#ifdef HAVE_UNITY
+#include <unity.h>
+#endif
+
 #define OPERATION_MINIMUM_TIME 2 /*s */
 #define NEEDS_ATTENTION_ANIMATION_TIMEOUT 2000 /*ms */
 #define REMOVE_FINISHED_OPERATIONS_TIEMOUT 3 /*s */
@@ -46,6 +52,9 @@ struct _NautilusProgressIndicator
     GtkPositionType popover_position;
 
     NautilusProgressInfoManager *progress_manager;
+#ifdef HAVE_UNITY
+    UnityLauncherEntry *ulauncher;
+#endif
 };
 
 G_DEFINE_FINAL_TYPE (NautilusProgressIndicator, nautilus_progress_indicator, ADW_TYPE_BIN);
@@ -116,6 +125,9 @@ on_remove_finished_operations_timeout (NautilusProgressIndicator *self)
     if (should_hide_operations_button (self))
     {
         nautilus_progress_indicator_set_reveal (self, FALSE);
+#ifdef HAVE_UNITY
+        unity_launcher_entry_set_progress_visible (self->ulauncher, FALSE);
+#endif
     }
     else
     {
@@ -257,6 +269,47 @@ get_paintable (GtkListItem          *listitem,
     return paintable;
 }
 
+#ifdef HAVE_UNITY
+static void
+update_unity_launcher_progress (NautilusProgressIndicator *self)
+{
+    gdouble total_work = 0.0;
+    gdouble finished_work = 0.0;
+    gdouble weighted_progress = 0.0;
+    guint not_started_items = 0;
+
+    guint n_items = g_list_model_get_n_items (G_LIST_MODEL (self->progress_infos_model));
+    for (size_t i = 0; i < n_items; i++)
+    {
+        NautilusProgressInfo *info = g_list_model_get_item (G_LIST_MODEL (self->progress_infos_model), i);
+        gdouble elapsed = nautilus_progress_info_get_total_elapsed_time (info);
+        gboolean has_finished = nautilus_progress_info_get_is_finished (info);
+        gdouble progress = has_finished ? 1.0 : nautilus_progress_info_get_progress (info);
+
+        if ((elapsed == 0.0) || (progress == 0.0))
+        {
+            not_started_items++;
+        }
+        else
+        {
+            finished_work += elapsed;
+
+            /* Estimate the total amount of time this task will take, based on
+             * how much time has already elapsed and its reported percentage progress.
+             */
+            total_work += elapsed / progress;
+        }
+    }
+
+    if (n_items == 0)
+        weighted_progress = 1.0;
+    else if (total_work > 0)
+        weighted_progress = finished_work * (n_items - not_started_items) / (total_work * n_items);
+
+    unity_launcher_entry_set_progress (self->ulauncher, weighted_progress);
+}
+#endif
+
 static void
 update_operations (NautilusProgressIndicator *self)
 {
@@ -274,6 +327,11 @@ update_operations (NautilusProgressIndicator *self)
     {
         add_operations_button_attention_style (self);
         nautilus_progress_indicator_set_reveal (self, TRUE);
+
+#ifdef HAVE_UNITY
+        update_unity_launcher_progress (self);
+        unity_launcher_entry_set_progress_visible (self->ulauncher, TRUE);
+#endif
     }
 
     /* Since we removed the info widgets, we need to restore the focus */
@@ -349,6 +407,11 @@ on_new_progress_info (NautilusProgressInfoManager *manager,
     g_signal_connect_object (info, "started",
                              G_CALLBACK (on_progress_info_started), self,
                              G_CONNECT_SWAPPED);
+#ifdef HAVE_UNITY
+    g_signal_connect_object (info, "progress-changed",
+                             G_CALLBACK (update_unity_launcher_progress), self,
+                             G_CONNECT_SWAPPED);
+#endif
     g_signal_connect_object (info, "finished",
                              G_CALLBACK (on_progress_info_finished), self,
                              G_CONNECT_SWAPPED);
@@ -482,6 +545,9 @@ nautilus_progress_indicator_constructed (GObject *object)
     NautilusProgressIndicator *self = NAUTILUS_PROGRESS_INDICATOR (object);
 
     self->progress_manager = nautilus_progress_info_manager_dup_singleton ();
+#ifdef HAVE_UNITY
+    self->ulauncher = unity_launcher_entry_get_for_desktop_id ("org.gnome.Nautilus.desktop");
+#endif
     g_signal_connect_object (self->progress_manager, "new-progress-info",
                              G_CALLBACK (on_new_progress_info), self,
                              G_CONNECT_DEFAULT);
diff --git a/src/unity-bookmarks-handler.c b/src/unity-bookmarks-handler.c
new file mode 100644
index 0000000..c8c3eb6
--- /dev/null
+++ b/src/unity-bookmarks-handler.c
@@ -0,0 +1,145 @@
+/*unity-bookmarks-handler.c: handle Unity bookmark for quicklist
+ *
+ * Copyright (C) 2012 Canonical
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors: Didier Roche <didrocks@ubuntu.com>
+ *
+ */
+
+#include <config.h>
+
+#include "unity-bookmarks-handler.h"
+
+#include <libdbusmenu-glib/dbusmenu-glib.h>
+#include "unity-quicklist-handler.h"
+
+#include "nautilus-application.h"
+#include "nautilus-bookmark.h"
+#include "nautilus-bookmark-list.h"
+
+#include <gdk/x11/gdkx.h>
+
+static UnityQuicklistHandler* unity_quicklist_handler = NULL;
+static NautilusBookmarkList* bookmarks = NULL;
+
+static void
+activate_bookmark_by_quicklist (DbusmenuMenuitem *menu,
+                                guint timestamp,
+                                NautilusBookmark *bookmark)
+{
+    GFile *location;
+    NautilusApplication *application;
+
+    g_return_if_fail (NAUTILUS_IS_BOOKMARK (bookmark));
+
+    location = nautilus_bookmark_get_location (bookmark);
+
+    application = NAUTILUS_APPLICATION (g_application_get_default ());
+
+    g_application_open (G_APPLICATION (application),
+                        (GFile* []) { location, NULL }, 1, "new-window");
+}
+
+static void
+unity_bookmarks_handler_remove_bookmark_quicklists (void)
+{
+
+    GList *children, *l;
+
+    /* remove unity quicklist bookmarks to launcher entries */
+    for (l = unity_quicklist_get_launcher_entries (unity_quicklist_handler); l; l = l->next)
+    {
+        UnityLauncherEntry *entry = l->data;
+        DbusmenuMenuitem *ql = unity_launcher_entry_get_quicklist (entry);
+        if (!ql)
+        {
+            break;
+        }
+
+        children = dbusmenu_menuitem_get_children (ql);
+        while (children)
+        {
+            DbusmenuMenuitem *child = children->data;
+            children = children->next;
+            if (unity_quicklist_handler_menuitem_is_bookmark_item (child))
+            {
+                g_signal_handlers_disconnect_matched (child, G_SIGNAL_MATCH_FUNC, 0, 0, 0,
+                                                      (GCallback) activate_bookmark_by_quicklist, 0);
+                dbusmenu_menuitem_child_delete (ql, child);
+                g_object_unref(child);
+            }
+        }
+    }
+}
+
+static void
+unity_bookmarks_handler_update_bookmarks (void)
+{
+    NautilusBookmark *bookmark;
+    GList *all_bookmarks;
+    GList *lb;
+    GList *l;
+    gchar *bn;
+    GString *bn_dbusmenu;
+    gchar *bookmark_name_dbusmenu;
+
+    /* append new set of bookmarks */
+    all_bookmarks = nautilus_bookmark_list_get_all (bookmarks);
+    for (lb = all_bookmarks; lb != NULL; lb = lb->next)
+    {
+        bookmark = NAUTILUS_BOOKMARK (lb->data);
+
+        for (l = unity_quicklist_get_launcher_entries (unity_quicklist_handler); l; l = l->next)
+        {
+            UnityLauncherEntry *entry = l->data;
+
+            DbusmenuMenuitem* menuitem = dbusmenu_menuitem_new();
+            bn = g_strdup(nautilus_bookmark_get_name (bookmark));
+            bn_dbusmenu = g_string_new_take (g_steal_pointer (&bn));
+            g_string_replace (bn_dbusmenu, "_", "__", 0);
+            bookmark_name_dbusmenu = g_string_free_and_steal (bn_dbusmenu);
+            dbusmenu_menuitem_property_set (menuitem, "label", bookmark_name_dbusmenu);
+            g_free (bookmark_name_dbusmenu);
+            g_signal_connect (menuitem, DBUSMENU_MENUITEM_SIGNAL_ITEM_ACTIVATED,
+                              (GCallback) activate_bookmark_by_quicklist,
+                              bookmark);
+
+            unity_quicklist_handler_append_menuitem (entry, menuitem);
+        }
+    }
+}
+
+static void
+unity_bookmarks_handler_refresh_bookmarks (void)
+{
+    unity_bookmarks_handler_remove_bookmark_quicklists ();
+    unity_bookmarks_handler_update_bookmarks ();
+}
+
+void
+unity_bookmarks_handler_initialize ()
+{
+    unity_quicklist_handler = unity_quicklist_handler_get_singleton ();
+    /* get the singleton */
+    bookmarks = nautilus_bookmark_list_new ();
+    unity_bookmarks_handler_refresh_bookmarks ();
+
+    /* Recreate dynamic part of menu if bookmark list changes */
+    g_signal_connect (bookmarks, "changed",
+                      G_CALLBACK (unity_bookmarks_handler_refresh_bookmarks), 0);
+}
diff --git a/extensions/image-properties/nautilus-image-properties-model.h b/src/unity-bookmarks-handler.h
similarity index 57%
copy from extensions/image-properties/nautilus-image-properties-model.h
copy to src/unity-bookmarks-handler.h
index 673c8e6..8257712 100644
--- a/extensions/image-properties/nautilus-image-properties-model.h
+++ b/src/unity-bookmarks-handler.h
@@ -1,4 +1,8 @@
-/* Copyright (C) 2004 Red Hat, Inc
+/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */
+/*
+ * unity-bookmarks-handler.h: handle Unity bookmark for quicklist
+ *
+ * Copyright (C) 2012 Canonical
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
@@ -11,13 +15,14 @@
  * General Public License for more details.
  *
  * You should have received a copy of the GNU General Public
- * License along with this program; if not, see <http://www.gnu.org/licenses/>.
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors: Didier Roche <didrocks@ubuntu.com>
  *
- * Author: Alexander Larsson <alexl@redhat.com>
  */
 
 #pragma once
 
-#include <nautilus-extension.h>
-
-NautilusPropertiesModel * nautilus_image_properties_model_new (NautilusFileInfo *file_info);
+void unity_bookmarks_handler_initialize (void);
diff --git a/src/unity-quicklist-handler.c b/src/unity-quicklist-handler.c
new file mode 100644
index 0000000..445921a
--- /dev/null
+++ b/src/unity-quicklist-handler.c
@@ -0,0 +1,166 @@
+/*unity-quicklist-handler.c: handle Unity quicklists
+ *
+ * Copyright (C) 2012 Canonical
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors: Didier Roche <didrocks@ubuntu.com>
+ *
+ */
+
+#include <config.h>
+
+#include "unity-quicklist-handler.h"
+
+struct _UnityQuicklistHandlerPriv {
+    GList *launcher_entries;
+};
+
+G_DEFINE_TYPE (UnityQuicklistHandler, unity_quicklist_handler, G_TYPE_OBJECT);
+
+static UnityQuicklistHandler *unity_quicklist_handler_singleton = NULL;
+
+GList *
+unity_quicklist_get_launcher_entries (UnityQuicklistHandler *self)
+{
+    return self->priv->launcher_entries;
+}
+
+gboolean
+unity_quicklist_handler_menuitem_is_progress_item (DbusmenuMenuitem *ql)
+{
+    g_return_val_if_fail(ql, FALSE);
+    return dbusmenu_menuitem_property_get_bool (ql, UNITY_QUICKLIST_PROGRESS_ITEM);
+}
+
+gboolean
+unity_quicklist_handler_menuitem_is_bookmark_item (DbusmenuMenuitem *ql)
+{
+    g_return_val_if_fail(ql, FALSE);
+    return (!unity_quicklist_handler_menuitem_is_progress_item(ql));
+}
+
+void
+unity_quicklist_handler_append_menuitem (UnityLauncherEntry *entry, DbusmenuMenuitem *elem)
+{
+    GList *children, *l;
+    int position;
+    DbusmenuMenuitem *ql;
+    gboolean is_bookmark;
+    gboolean is_progress;
+
+    g_return_if_fail (entry);
+
+    position = 0;
+    ql = unity_launcher_entry_get_quicklist (entry);
+
+    is_bookmark = unity_quicklist_handler_menuitem_is_bookmark_item (elem);
+    is_progress = unity_quicklist_handler_menuitem_is_progress_item (elem);
+
+    if (!ql)
+    {
+        ql = dbusmenu_menuitem_new ();
+        unity_launcher_entry_set_quicklist (entry, ql);
+    }
+
+    children = dbusmenu_menuitem_get_children (ql);
+    for (l = children; l; l = l->next)
+    {
+        DbusmenuMenuitem *child = l->data;
+        /* set quicklist groups together, and bookmarks group after progress group.
+           bookmarks elements are ordered alphabetically */
+        if ((is_bookmark && unity_quicklist_handler_menuitem_is_bookmark_item (child) &&
+                (g_strcmp0 (dbusmenu_menuitem_property_get (child, DBUSMENU_MENUITEM_PROP_LABEL), dbusmenu_menuitem_property_get (elem, DBUSMENU_MENUITEM_PROP_LABEL)) < 0)) ||
+            (is_progress && unity_quicklist_handler_menuitem_is_progress_item (child)) ||
+            (is_progress && unity_quicklist_handler_menuitem_is_bookmark_item (child)))
+        {
+            position++;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    dbusmenu_menuitem_child_add_position (ql, elem, position);
+}
+
+static void
+unity_quicklist_handler_dispose (GObject *obj)
+{
+    UnityQuicklistHandler *self = UNITY_QUICKLIST_HANDLER (obj);
+
+    if (self->priv->launcher_entries)
+    {
+        g_list_free_full (self->priv->launcher_entries, g_object_unref);
+        self->priv->launcher_entries = NULL;
+    }
+
+    G_OBJECT_CLASS (unity_quicklist_handler_parent_class)->dispose (obj);
+}
+
+static void
+unity_quicklist_handler_launcher_entry_add (UnityQuicklistHandler *self,
+                                            const gchar *entry_id)
+{
+    GList **entries;
+    UnityLauncherEntry *entry;
+
+    entries = &(self->priv->launcher_entries);
+    entry = unity_launcher_entry_get_for_desktop_id (entry_id);
+
+    if (entry)
+    {
+        *entries = g_list_prepend (*entries, entry);
+    }
+}
+
+static void
+unity_quicklist_handler_init (UnityQuicklistHandler *self)
+{
+    self->priv = G_TYPE_INSTANCE_GET_PRIVATE (self, UNITY_TYPE_QUICKLIST_HANDLER,
+                                              UnityQuicklistHandlerPriv);
+
+    unity_quicklist_handler_launcher_entry_add (self, "org.gnome.Nautilus.desktop");
+    g_return_if_fail (g_list_length (self->priv->launcher_entries) != 0);
+}
+
+static void
+unity_quicklist_handler_class_init (UnityQuicklistHandlerClass *klass)
+{
+    GObjectClass *oclass;
+
+    oclass = G_OBJECT_CLASS (klass);
+    oclass->dispose = unity_quicklist_handler_dispose;
+
+    g_type_class_add_private (klass, sizeof (UnityQuicklistHandlerPriv));
+}
+
+UnityQuicklistHandler *
+unity_quicklist_handler_get_singleton (void)
+{
+    if (!unity_quicklist_handler_singleton)
+    {
+        unity_quicklist_handler_singleton = unity_quicklist_handler_new ();
+    }
+    return unity_quicklist_handler_singleton;
+}
+
+UnityQuicklistHandler *
+unity_quicklist_handler_new (void)
+{
+    return g_object_new (UNITY_TYPE_QUICKLIST_HANDLER, NULL);
+}
diff --git a/src/unity-quicklist-handler.h b/src/unity-quicklist-handler.h
new file mode 100644
index 0000000..89893d5
--- /dev/null
+++ b/src/unity-quicklist-handler.h
@@ -0,0 +1,69 @@
+/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */
+/*
+ * unity-quicklist.h: handle unity quicklists.
+ *
+ * Copyright (C) 2012 Canonical
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors: Didier Roche <didrocks@ubuntu.com>
+ *
+ */
+
+#pragma once
+
+#include <glib-object.h>
+#include <glib/gi18n.h>
+
+#include <libdbusmenu-glib/dbusmenu-glib.h>
+#include <unity.h>
+
+#define UNITY_TYPE_QUICKLIST_HANDLER unity_quicklist_handler_get_type()
+#define UNITY_QUICKLIST_HANDLER(obj) \
+  (G_TYPE_CHECK_INSTANCE_CAST ((obj), UNITY_TYPE_QUICKLIST_HANDLER, UnityQuicklistHandler))
+#define UNITY_QUICKLIST_HANDLER_CLASS(klass) \
+  (G_TYPE_CHECK_CLASS_CAST ((klass), UNITY_TYPE_QUICKLIST_HANDLER, UnityQuicklistHandlerClass))
+#define UNITY_IS_QUICKLIST_HANDLER(obj) \
+  (G_TYPE_CHECK_INSTANCE_TYPE ((obj), UNITY_TYPE_QUICKLIST_HANDLER))
+#define UNITY_IS_QUICKLIST_HANDLER_CLASS(klass) \
+  (G_TYPE_CHECK_CLASS_TYPE ((klass), UNITY_TYPE_QUICKLIST_HANDLER))
+#define UNITY_QUICKLIST_HANDLER_GET_CLASS(obj) \
+  (G_TYPE_INSTANCE_GET_CLASS ((obj), UNITY_TYPE_QUICKLIST_HANDLER, UnityQuicklistHandlerClass))
+
+typedef struct _UnityQuicklistHandlerPriv UnityQuicklistHandlerPriv;
+
+#define UNITY_QUICKLIST_PROGRESS_ITEM "unity-quicklist-progress-item"
+
+typedef struct {
+  GObject parent;
+
+  /* private */
+  UnityQuicklistHandlerPriv *priv;
+} UnityQuicklistHandler;
+
+typedef struct {
+  GObjectClass parent_class;
+} UnityQuicklistHandlerClass;
+
+GType unity_quicklist_handler_get_type (void);
+
+UnityQuicklistHandler * unity_quicklist_handler_new (void);
+UnityQuicklistHandler * unity_quicklist_handler_get_singleton (void);
+
+GList * unity_quicklist_get_launcher_entries (UnityQuicklistHandler *unity_quicklist_handler);
+gboolean unity_quicklist_handler_menuitem_is_progress_item (DbusmenuMenuitem *ql);
+gboolean unity_quicklist_handler_menuitem_is_bookmark_item (DbusmenuMenuitem *ql);
+void unity_quicklist_handler_append_menuitem (UnityLauncherEntry *entry, DbusmenuMenuitem *elem);
